server/service/src/sensor/berlinger.rs

```diff
@@ -339,33 +331,440 @@ pub fn read_sensor(
     };
 
     for temperature_sensor_log in temperature_sensor_logs {
-        sensor_add_log_if_new(connection, &record.sensor_row, &temperature_sensor_log)?;
+        sensor_add_log_if_new(connection, &sensor_row, &temperature_sensor_log)?;
     }
 
-    for temperature_sensor_breach in temperature_sensor_breaches {
+    // Add consequitive then cumulative breaches, order is important because breach and log association
+    // is priorities for consequitive breach i.e. if log is in both cumulative and consequitive breach
+    // the breach id would be from consequitive
+    for temperature_sensor_breach in sort_breaches_by_type(temperature_sensor_breaches) {
         // Look up matching config from the USB data and snapshot it as part of the breach
         if let Some(temperature_sensor_config) = temperature_sensor_configs
             .iter()
             .find(|&t| t.breach_type == temperature_sensor_breach.breach_type)
         {
-            sensor_add_breach_if_new(
+            let upserted_breach = sensor_add_breach_if_new(
                 connection,
-                &record.sensor_row,
+                &sensor_row,
                 &temperature_sensor_breach,
                 &temperature_sensor_config,
             )?;
+
+            if let Some(upserted_breach) = upserted_breach {
+                update_sensor_logs_for_breach(connection, &upserted_breach)?;
+            }
         }
     }
 
     // Finally, update sensor's last connected time if it has changed
-    if record.sensor_row.last_connection_datetime != temperature_sensor.last_connected_timestamp {
+    if sensor_row.last_connection_datetime != temperature_sensor.last_connected_timestamp {
         SensorRowRepository::new(connection).upsert_one(&SensorRow {
             last_connection_datetime: temperature_sensor.last_connected_timestamp,
-            ..record.sensor_row
+            ..sensor_row
         })?;
     }
 
     Ok(result)
 }
 
-// TODO tests
+// First of all consequitive and then cumulative
+fn breach_sort_weight(breach: &TemperatureBreachRowType) -> u8 {
+    use TemperatureBreachRowType::*;
+    match breach {
+        ColdConsecutive => 1,
+        HotConsecutive => 2,
+        ColdCumulative => 3,
+        HotCumulative => 4,
+    }
+}
+
+fn sort_breaches_by_type(
+    mut breaches: Vec<temperature_sensor::TemperatureBreach>,
+) -> Vec<temperature_sensor::TemperatureBreach> {
+    breaches.sort_by(|a, b| {
+        breach_sort_weight(&get_breach_row_type(&a.breach_type))
+            .cmp(&breach_sort_weight(&get_breach_row_type(&b.breach_type)))
+    });
+
+    breaches
+}
+
+#[cfg(test)]
+mod test {
```

Comment:
Pretty happy with the test, two steps, one insert and second one insert more logs, updates existing breach and updates existing logs with breach_id